packages/frontend-core/src/components/grid/controls/ColumnsSettingContent.svelte

```diff
@@ -1,87 +1,179 @@
 <script>
   import { getContext } from "svelte"
-  import { Icon, notifications } from "@budibase/bbui"
+  import { Icon, notifications, ActionButton, Popover } from "@budibase/bbui"
   import { getColumnIcon } from "../lib/utils"
   import ToggleActionButtonGroup from "./ToggleActionButtonGroup.svelte"
   import { helpers } from "@budibase/shared-core"
+  import { FieldType } from "@budibase/types"
+  import { FieldPermissions } from "../../../constants"
 
-  export let allowViewReadonlyColumns = false
+  export let permissions = [FieldPermissions.WRITABLE, FieldPermissions.HIDDEN]
+  export let disabledPermissions = []
+  export let columns
+  export let fromRelationshipField
 
-  const { columns, datasource, dispatch } = getContext("grid")
+  const { datasource, dispatch, cache } = getContext("grid")
 
-  const toggleColumn = async (column, permission) => {
-    const visible = permission !== PERMISSION_OPTIONS.HIDDEN
-    const readonly = permission === PERMISSION_OPTIONS.READONLY
+  let relationshipPanelAnchor
+  let relationshipFieldName
 
-    await datasource.actions.addSchemaMutation(column.name, {
-      visible,
-      readonly,
-    })
-    try {
-      await datasource.actions.saveSchemaMutations()
-    } catch (e) {
-      notifications.error(e.message)
-    } finally {
-      await datasource.actions.resetSchemaMutations()
-      await datasource.actions.refreshDefinition()
+  $: relationshipField = columns.find(
+    c => c.name === relationshipFieldName
+  )?.schema
+  $: permissionsObj = permissions.reduce(
+    (acc, c) => ({
+      ...acc,
+      [c]: {
+        disabled: disabledPermissions.includes(c),
+      },
+    }),
+    {}
+  )
+
+  $: allowRelationshipSchemas = true // TODO
+
+  $: displayColumns = columns.map(c => {
```

Comment:
could we break this out into a function? Right now I think this will react to a lot of different possible changes - maybe that is desired, but I'm guessing it should only react if the columns change

Reply:
Most of these things will be quite static, but we depend on reacting to all of them if loaded async or something changes via its parameters. For this it was created like this, otherwise it would need a function that gets almost all the same parameters